src/QsCompiler/QirGeneration/Context.cs

```diff
@@ -1268,6 +1268,30 @@ Value ProcessConditionalBlock(BasicBlock block, Func<IValue>? evaluate)
                 }
 
                 this.SetCurrentBlock(contBlock);
+
+                // Handle special cases with constant values where the phi node is not necessary.
+                if (QirValues.AsConstantUInt32(evaluatedOnTrue) == 0 && evaluatedOnFalse == condition)
+                {
+                    // This case is the equivalent of: `val = cond ? false : cond;`
+                    // which is the same as: `val = false;` so we can just return the `false` value
```

Comment:
But you may not have executed the onTrue block... So no, I don't think you can return the evaluatedOnTrue value.

Reply:
Do you mean executed at runtime or generated at compile time?
For both `evaluatedOnTrue` and `evaluatedOnFalse` they are are always processed by generation if a lambda was provided for them, otherwise they just get the default value. The value they represent doesn't come from the true/false block, but from the block that produces the conditional value itself. That's why I can know at this point in the generation code if they are static Boolean values and optimize appropriately.

The more generic notion of this optimization is one other languages warn about: using an if-statement to produce a Boolean value is inefficient and can be replaced by the condition in the if-statement itself or a constant. So what I'm doing here is the equivalent of optimizing:
```
bool val;
if cond {
    val = false;
} else {
    val = cond;
}
```
into:
```
bool val = false;
```

Reply:
Alright that works - it is a bit confusing to read for me (so I would have instead explicitly created a new constant for the sake of readability), but I get what you are saying.